File: sdks/strat-sdk.js
/**
 * STRAT JavaScript SDK
 * Complete API wrapper for STRAT blockchain
 * @version 1.0.0
 */

const axios = require('axios');
const EventEmitter = require('events');

class StratSDK extends EventEmitter {
  constructor(config = {}) {
    super();
    this.apiUrl = config.apiUrl || 'http://localhost:3000';
    this.wsUrl = config.wsUrl || 'ws://localhost:3000';
    this.apiKey = config.apiKey || null;
    this.timeout = config.timeout || 30000;
    this.ws = null;
  }

  // HTTP Request Handler
  async request(method, endpoint, data = null, options = {}) {
    try {
      const config = {
        method,
        url: `${this.apiUrl}${endpoint}`,
        timeout: this.timeout,
        headers: {
          'Content-Type': 'application/json',
          ...(this.apiKey && { 'Authorization': `Bearer ${this.apiKey}` })
        }
      };

      if (data) {
        if (method === 'GET') {
          config.params = data;
        } else {
          config.data = data;
        }
      }

      const response = await axios(config);
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  handleError(error) {
    if (error.response) {
      return new Error(`API Error (${error.response.status}): ${error.response.data.error || error.response.statusText}`);
    } else if (error.request) {
      return new Error(`Network Error: Could not connect to ${this.apiUrl}`);
    } else {
      return new Error(`Error: ${error.message}`);
    }
  }

  // Blockchain Methods
  async getBlockchainInfo() {
    return await this.request('GET', '/api/blockchain/info');
  }

  async getBlock(identifier) {
    const endpoint = isNaN(identifier)
      ? `/api/blockchain/block/${identifier}`
      : `/api/blockchain/block-by-index/${identifier}`;
    return await this.request('GET', endpoint);
  }

  async getLatestBlocks(count = 10) {
    return await this.request('GET', '/api/blockchain/blocks', { limit: count });
  }

  async getBlockByHash(hash) {
    return await this.request('GET', `/api/blockchain/block/${hash}`);
  }

  async getBlockByIndex(index) {
    return await this.request('GET', `/api/blockchain/block-by-index/${index}`);
  }

  // Transaction Methods
  async getTransaction(txId) {
    return await this.request('GET', `/api/transactions/${txId}`);
  }

  async sendTransaction(fromAddress, toAddress, amount, privateKey) {
    return await this.request('POST', '/api/transactions/send', {
      fromAddress,
      toAddress,
      amount,
      privateKey
    });
  }

  async getTransactionHistory(address, options = {}) {
    return await this.request('GET', `/api/transactions/history/${address}`, options);
  }

  async getPendingTransactions() {
    return await this.request('GET', '/api/transactions/pending');
  }

  // Wallet Methods
  async createWallet(username, password) {
    return await this.request('POST', '/api/auth/register', { username, password });
  }

  async getBalance(address) {
    const result = await this.request('GET', `/api/wallets/balance/${address}`);
    return result.balance;
  }

  async getWalletInfo(address) {
    return await this.request('GET', `/api/wallets/${address}`);
  }

  async getUTXOs(address) {
    return await this.request('GET', `/api/wallets/utxos/${address}`);
  }

  // Smart Contract Methods
  async deployContract(code, owner, privateKey) {
    return await this.request('POST', '/api/contracts/deploy', {
      code,
      owner,
      privateKey
    });
  }

  async callContract(contractAddress, method, params, caller, privateKey) {
    return await this.request('POST', '/api/contracts/call', {
      contractAddress,
      method,
      params,
      caller,
      privateKey
    });
  }

  async getContract(contractAddress) {
    return await this.request('GET', `/api/contracts/${contractAddress}`);
  }

  async getContractState(contractAddress) {
    return await this.request('GET', `/api/contracts/${contractAddress}/state`);
  }

  async listContracts(options = {}) {
    return await this.request('GET', '/api/contracts', options);
  }

  // Mining Methods
  async getMiningInfo() {
    return await this.request('GET', '/api/mining/info');
  }

  async startMining(minerAddress) {
    return await this.request('POST', '/api/mining/start', { minerAddress });
  }

  async stopMining() {
    return await this.request('POST', '/api/mining/stop');
  }

  async getMiningStats(address) {
    return await this.request('GET', `/api/mining/stats/${address}`);
  }

  // Mempool Methods
  async getMempoolInfo() {
    return await this.request('GET', '/api/mempool/stats');
  }

  async getMempoolTransactions() {
    return await this.request('GET', '/api/mempool/transactions');
  }

  // Staking Methods
  async stake(address, amount, privateKey) {
    return await this.request('POST', '/api/staking/stake', {
      address,
      amount,
      privateKey
    });
  }

  async unstake(address, amount, privateKey) {
    return await this.request('POST', '/api/staking/unstake', {
      address,
      amount,
      privateKey
    });
  }

  async getStakingInfo(address) {
    return await this.request('GET', `/api/staking/info/${address}`);
  }

  async claimRewards(address, privateKey) {
    return await this.request('POST', '/api/staking/claim', {
      address,
      privateKey
    });
  }

  // NFT Methods
  async mintNFT(data) {
    return await this.request('POST', '/api/nft/mint', data);
  }

  async getNFT(tokenId) {
    return await this.request('GET', `/api/nft/${tokenId}`);
  }

  async transferNFT(tokenId, from, to, privateKey) {
    return await this.request('POST', '/api/nft/transfer', {
      tokenId,
      from,
      to,
      privateKey
    });
  }

  async listNFTs(owner) {
    return await this.request('GET', `/api/nft/list/${owner}`);
  }

  // Social/Governance Methods
  async createProposal(data) {
    return await this.request('POST', '/api/governance/proposal', data);
  }

  async vote(proposalId, vote, voter, privateKey) {
    return await this.request('POST', '/api/governance/vote', {
      proposalId,
      vote,
      voter,
      privateKey
    });
  }

  async getProposal(proposalId) {
    return await this.request('GET', `/api/governance/proposal/${proposalId}`);
  }

  async listProposals(options = {}) {
    return await this.request('GET', '/api/governance/proposals', options);
  }

  // Explorer Methods
  async searchAddress(address) {
    return await this.request('GET', `/api/explorer/address/${address}`);
  }

  async getRichList(limit = 100) {
    return await this.request('GET', '/api/explorer/richlist', { limit });
  }

  async getNetworkStats() {
    return await this.request('GET', '/api/explorer/stats');
  }

  // WebSocket Methods
  connectWebSocket() {
    if (typeof window !== 'undefined' && window.io) {
      // Browser environment
      this.ws = window.io(this.wsUrl);
    } else {
      // Node.js environment
      const io = require('socket.io-client');
      this.ws = io(this.wsUrl);
    }

    this.ws.on('connect', () => {
      this.emit('connected');
    });

    this.ws.on('disconnect', () => {
      this.emit('disconnected');
    });

    this.ws.on('stats', (data) => {
      this.emit('stats', data);
    });

    this.ws.on('new_block', (block) => {
      this.emit('newBlock', block);
    });

    this.ws.on('new_transaction', (tx) => {
      this.emit('newTransaction', tx);
    });

    this.ws.on('mempool_stats', (stats) => {
      this.emit('mempoolStats', stats);
    });

    this.ws.on('address_balance', (data) => {
      this.emit('balanceUpdate', data);
    });

    return this.ws;
  }

  subscribeToAddress(address) {
    if (!this.ws) throw new Error('WebSocket not connected');
    this.ws.emit('subscribe_address', address);
  }

  unsubscribeFromAddress(address) {
    if (!this.ws) throw new Error('WebSocket not connected');
    this.ws.emit('unsubscribe_address', address);
  }

  subscribeToBlocks() {
    if (!this.ws) throw new Error('WebSocket not connected');
    this.ws.emit('subscribe_blocks');
  }

  unsubscribeFromBlocks() {
    if (!this.ws) throw new Error('WebSocket not connected');
    this.ws.emit('unsubscribe_blocks');
  }

  subscribeToMempool() {
    if (!this.ws) throw new Error('WebSocket not connected');
    this.ws.emit('subscribe_mempool');
  }

  unsubscribeFromMempool() {
    if (!this.ws) throw new Error('WebSocket not connected');
    this.ws.emit('unsubscribe_mempool');
  }

  disconnectWebSocket() {
    if (this.ws) {
      this.ws.disconnect();
      this.ws = null;
    }
  }

  // Utility Methods
  async healthCheck() {
    return await this.request('GET', '/health');
  }

  async getAPIVersion() {
    return await this.request('GET', '/api');
  }

  // Batch Operations
  async batchRequest(requests) {
    const promises = requests.map(req =>
      this.request(req.method, req.endpoint, req.data)
    );

    return await Promise.allSettled(promises);
  }

  // Transaction Builder Helper
  buildTransaction(from, to, amount) {
    return {
      fromAddress: from,
      toAddress: to,
      amount: parseFloat(amount),
      timestamp: Date.now()
    };
  }

  // Address Validation
  isValidAddress(address) {
    return /^0x[a-fA-F0-9]{40}$/.test(address);
  }

  // Amount Conversion
  toWei(amount) {
    return amount * 1e18;
  }

  fromWei(amount) {
    return amount / 1e18;
  }
}

// Browser and Node.js compatibility
if (typeof module !== 'undefined' && module.exports) {
  module.exports = StratSDK;
}

if (typeof window !== 'undefined') {
  window.StratSDK = StratSDK;
}

File: sdks/strat_sdk.rs
// STRAT Rust SDK
// Complete API wrapper for STRAT blockchain
// Version 1.0.0

use reqwest::{Client, Error as ReqwestError};
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::collections::HashMap;
use std::time::Duration;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum StratError {
    #[error("Network error: {0}")]
    NetworkError(String),

    #[error("API error ({status}): {message}")]
    ApiError { status: u16, message: String },

    #[error("Serialization error: {0}")]
    SerializationError(String),

    #[error("Invalid address format")]
    InvalidAddress,
}

impl From<ReqwestError> for StratError {
    fn from(err: ReqwestError) -> Self {
        StratError::NetworkError(err.to_string())
    }
}

impl From<serde_json::Error> for StratError {
    fn from(err: serde_json::Error) -> Self {
        StratError::SerializationError(err.to_string())
    }
}

#[derive(Debug, Clone)]
pub struct StratSDKConfig {
    pub api_url: String,
    pub api_key: Option<String>,
    pub timeout: Duration,
}

impl Default for StratSDKConfig {
    fn default() -> Self {
        Self {
            api_url: "http://localhost:3000".to_string(),
            api_key: None,
            timeout: Duration::from_secs(30),
        }
    }
}

pub struct StratSDK {
    config: StratSDKConfig,
    client: Client,
}

impl StratSDK {
    pub fn new(config: StratSDKConfig) -> Result<Self, StratError> {
        let client = Client::builder()
            .timeout(config.timeout)
            .build()?;

        Ok(Self { config, client })
    }

    pub fn with_default() -> Result<Self, StratError> {
        Self::new(StratSDKConfig::default())
    }

    async fn request(
        &self,
        method: reqwest::Method,
        endpoint: &str,
        body: Option<Value>,
    ) -> Result<Value, StratError> {
        let url = format!("{}{}", self.config.api_url, endpoint);

        let mut request = self.client.request(method, &url);

        if let Some(api_key) = &self.config.api_key {
            request = request.header("Authorization", format!("Bearer {}", api_key));
        }

        if let Some(body) = body {
            request = request.json(&body);
        }

        let response = request.send().await?;
        let status = response.status();

        if status.is_success() {
            let data: Value = response.json().await?;
            Ok(data)
        } else {
            let error_data: Value = response.json().await?;
            let message = error_data["error"]
                .as_str()
                .unwrap_or("Unknown error")
                .to_string();

            Err(StratError::ApiError {
                status: status.as_u16(),
                message,
            })
        }
    }

    // Blockchain Methods

    pub async fn get_blockchain_info(&self) -> Result<Value, StratError> {
        self.request(reqwest::Method::GET, "/api/blockchain/info", None)
            .await
    }

    pub async fn get_block(&self, identifier: &str) -> Result<Value, StratError> {
        let endpoint = if identifier.chars().all(|c| c.is_numeric()) {
            format!("/api/blockchain/block-by-index/{}", identifier)
        } else {
            format!("/api/blockchain/block/{}", identifier)
        };

        self.request(reqwest::Method::GET, &endpoint, None).await
    }

    pub async fn get_latest_blocks(&self, count: u32) -> Result<Value, StratError> {
        self.request(
            reqwest::Method::GET,
            &format!("/api/blockchain/blocks?limit={}", count),
            None,
        )
        .await
    }

    // Transaction Methods

    pub async fn get_transaction(&self, tx_id: &str) -> Result<Value, StratError> {
        self.request(
            reqwest::Method::GET,
            &format!("/api/transactions/{}", tx_id),
            None,
        )
        .await
    }

    pub async fn send_transaction(
        &self,
        from_address: &str,
        to_address: &str,
        amount: f64,
        private_key: &str,
    ) -> Result<Value, StratError> {
        let body = json!({
            "fromAddress": from_address,
            "toAddress": to_address,
            "amount": amount,
            "privateKey": private_key,
        });

        self.request(reqwest::Method::POST, "/api/transactions/send", Some(body))
            .await
    }

    pub async fn get_transaction_history(
        &self,
        address: &str,
        limit: u32,
        offset: u32,
    ) -> Result<Value, StratError> {
        self.request(
            reqwest::Method::GET,
            &format!(
                "/api/transactions/history/{}?limit={}&offset={}",
                address, limit, offset
            ),
            None,
        )
        .await
    }

    pub async fn get_pending_transactions(&self) -> Result<Value, StratError> {
        self.request(reqwest::Method::GET, "/api/transactions/pending", None)
            .await
    }

    // Wallet Methods

    pub async fn create_wallet(&self, username: &str, password: &str) -> Result<Value, StratError> {
        let body = json!({
            "username": username,
            "password": password,
        });

        self.request(reqwest::Method::POST, "/api/auth/register", Some(body))
            .await
    }

    pub async fn get_balance(&self, address: &str) -> Result<f64, StratError> {
        let result = self
            .request(
                reqwest::Method::GET,
                &format!("/api/wallets/balance/{}", address),
                None,
            )
            .await?;

        result["balance"]
            .as_f64()
            .ok_or_else(|| StratError::SerializationError("Invalid balance format".to_string()))
    }

    pub async fn get_wallet_info(&self, address: &str) -> Result<Value, StratError> {
        self.request(
            reqwest::Method::GET,
            &format!("/api/wallets/{}", address),
            None,
        )
        .await
    }

    pub async fn get_utxos(&self, address: &str) -> Result<Value, StratError> {
        self.request(
            reqwest::Method::GET,
            &format!("/api/wallets/utxos/{}", address),
            None,
        )
        .await
    }

    // Smart Contract Methods

    pub async fn deploy_contract(
        &self,
        code: &str,
        owner: &str,
        private_key: &str,
    ) -> Result<Value, StratError> {
        let body = json!({
            "code": code,
            "owner": owner,
            "privateKey": private_key,
        });

        self.request(reqwest::Method::POST, "/api/contracts/deploy", Some(body))
            .await
    }

    pub async fn call_contract(
        &self,
        contract_address: &str,
        method: &str,
        params: Vec<Value>,
        caller: &str,
        private_key: &str,
    ) -> Result<Value, StratError> {
        let body = json!({
            "contractAddress": contract_address,
            "method": method,
            "params": params,
            "caller": caller,
            "privateKey": private_key,
        });

        self.request(reqwest::Method::POST, "/api/contracts/call", Some(body))
            .await
    }

    pub async fn get_contract(&self, contract_address: &str) -> Result<Value, StratError> {
        self.request(
            reqwest::Method::GET,
            &format!("/api/contracts/{}", contract_address),
            None,
        )
        .await
    }

    pub async fn get_contract_state(&self, contract_address: &str) -> Result<Value, StratError> {
        self.request(
            reqwest::Method::GET,
            &format!("/api/contracts/{}/state", contract_address),
            None,
        )
        .await
    }

    // Mining Methods

    pub async fn get_mining_info(&self) -> Result<Value, StratError> {
        self.request(reqwest::Method::GET, "/api/mining/info", None)
            .await
    }

    pub async fn start_mining(&self, miner_address: &str) -> Result<Value, StratError> {
        let body = json!({
            "minerAddress": miner_address,
        });

        self.request(reqwest::Method::POST, "/api/mining/start", Some(body))
            .await
    }

    pub async fn stop_mining(&self) -> Result<Value, StratError> {
        self.request(reqwest::Method::POST, "/api/mining/stop", None)
            .await
    }

    pub async fn get_mining_stats(&self, address: &str) -> Result<Value, StratError> {
        self.request(
            reqwest::Method::GET,
            &format!("/api/mining/stats/{}", address),
            None,
        )
        .await
    }

    // Staking Methods

    pub async fn stake(
        &self,
        address: &str,
        amount: f64,
        private_key: &str,
    ) -> Result<Value, StratError> {
        let body = json!({
            "address": address,
            "amount": amount,
            "privateKey": private_key,
        });

        self.request(reqwest::Method::POST, "/api/staking/stake", Some(body))
            .await
    }

    pub async fn unstake(
        &self,
        address: &str,
        amount: f64,
        private_key: &str,
    ) -> Result<Value, StratError> {
        let body = json!({
            "address": address,
            "amount": amount,
            "privateKey": private_key,
        });

        self.request(reqwest::Method::POST, "/api/staking/unstake", Some(body))
            .await
    }

    pub async fn get_staking_info(&self, address: &str) -> Result<Value, StratError> {
        self.request(
            reqwest::Method::GET,
            &format!("/api/staking/info/{}", address),
            None,
        )
        .await
    }

    pub async fn claim_rewards(&self, address: &str, private_key: &str) -> Result<Value, StratError> {
        let body = json!({
            "address": address,
            "privateKey": private_key,
        });

        self.request(reqwest::Method::POST, "/api/staking/claim", Some(body))
            .await
    }

    // NFT Methods

    pub async fn mint_nft(&self, data: Value) -> Result<Value, StratError> {
        self.request(reqwest::Method::POST, "/api/nft/mint", Some(data))
            .await
    }

    pub async fn get_nft(&self, token_id: u64) -> Result<Value, StratError> {
        self.request(
            reqwest::Method::GET,
            &format!("/api/nft/{}", token_id),
            None,
        )
        .await
    }

    pub async fn transfer_nft(
        &self,
        token_id: u64,
        from: &str,
        to: &str,
        private_key: &str,
    ) -> Result<Value, StratError> {
        let body = json!({
            "tokenId": token_id,
            "from": from,
            "to": to,
            "privateKey": private_key,
        });

        self.request(reqwest::Method::POST, "/api/nft/transfer", Some(body))
            .await
    }

    pub async fn list_nfts(&self, owner: &str) -> Result<Value, StratError> {
        self.request(
            reqwest::Method::GET,
            &format!("/api/nft/list/{}", owner),
            None,
        )
        .await
    }

    // Governance Methods

    pub async fn create_proposal(&self, data: Value) -> Result<Value, StratError> {
        self.request(reqwest::Method::POST, "/api/governance/proposal", Some(data))
            .await
    }

    pub async fn vote(
        &self,
        proposal_id: u64,
        vote: bool,
        voter: &str,
        private_key: &str,
    ) -> Result<Value, StratError> {
        let body = json!({
            "proposalId": proposal_id,
            "vote": vote,
            "voter": voter,
            "privateKey": private_key,
        });

        self.request(reqwest::Method::POST, "/api/governance/vote", Some(body))
            .await
    }

    pub async fn get_proposal(&self, proposal_id: u64) -> Result<Value, StratError> {
        self.request(
            reqwest::Method::GET,
            &format!("/api/governance/proposal/{}", proposal_id),
            None,
        )
        .await
    }

    pub async fn list_proposals(&self) -> Result<Value, StratError> {
        self.request(reqwest::Method::GET, "/api/governance/proposals", None)
            .await
    }

    // Explorer Methods

    pub async fn search_address(&self, address: &str) -> Result<Value, StratError> {
        self.request(
            reqwest::Method::GET,
            &format!("/api/explorer/address/{}", address),
            None,
        )
        .await
    }

    pub async fn get_rich_list(&self, limit: u32) -> Result<Value, StratError> {
        self.request(
            reqwest::Method::GET,
            &format!("/api/explorer/richlist?limit={}", limit),
            None,
        )
        .await
    }

    pub async fn get_network_stats(&self) -> Result<Value, StratError> {
        self.request(reqwest::Method::GET, "/api/explorer/stats", None)
            .await
    }

    // Utility Methods

    pub async fn health_check(&self) -> Result<Value, StratError> {
        self.request(reqwest::Method::GET, "/health", None).await
    }

    pub fn is_valid_address(address: &str) -> bool {
        let re = regex::Regex::new(r"^0x[a-fA-F0-9]{40}$").unwrap();
        re.is_match(address)
    }

    pub fn to_wei(amount: f64) -> i64 {
        (amount * 1e18) as i64
    }

    pub fn from_wei(wei: i64) -> f64 {
        wei as f64 / 1e18
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_sdk_creation() {
        let sdk = StratSDK::with_default();
        assert!(sdk.is_ok());
    }

    #[test]
    fn test_address_validation() {
        assert!(StratSDK::is_valid_address(
            "0x1234567890abcdef1234567890abcdef12345678"
        ));
        assert!(!StratSDK::is_valid_address("invalid"));
    }

    #[test]
    fn test_wei_conversion() {
        assert_eq!(StratSDK::to_wei(1.0), 1000000000000000000);
        assert_eq!(StratSDK::from_wei(1000000000000000000), 1.0);
    }
}
